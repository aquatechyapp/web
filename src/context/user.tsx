import { LoadingSpinner } from '../components/LoadingSpinner';
import { User } from '../interfaces/User';
import { clientAxios } from '../lib/clientAxios';
import { isEmpty } from '../utils';
import { useQuery, useQueryClient } from '@tanstack/react-query';
import Cookies from 'js-cookie';
import { usePathname, useRouter } from 'next/navigation';
import { createContext, useContext, useEffect, useState } from 'react';

type UserContextType = {
  user: User | undefined;
  setUser: (user: User) => void;
};

const UserContext = createContext<UserContextType>({
  user: undefined,
  setUser: () => {}
});

export const UserProvider = ({ children }: { children: React.ReactNode }) => {
  const [user, setUser] = useState<User>();
  const userId = Cookies.get('userId');
  const { push } = useRouter();
  const pathname = usePathname();
  const queryClient = useQueryClient();

  const { data, isLoading, isError } = useQuery({
    queryKey: ['user', userId],
    queryFn: async () => {
      if (pathname === '/login' || pathname === '/signup') {
        queryClient.cancelQueries({ queryKey: ['user'] });
        return {};
      }
      if (!userId) {
        queryClient.cancelQueries({ queryKey: ['user'] });
        push('/login');
        return {};
      }

      const response = await clientAxios.get(`/users/${userId}`);

      const user = {
        ...response.data.user,
        incomeAsACompany: response.data.incomeAsACompany,
        incomeAsASubcontractor: response.data.incomeAsASubcontractor
      };

      setUser(user);

      return user;
    }
  });

  useEffect(() => {
    if (isEmpty(data)) return;
    setUser(data);
  }, [data, userId]);

  if (isLoading) return <LoadingSpinner />;
  if (isError) return push('/login');

  return (
    <UserContext.Provider value={{ user, setUser }}>
      {children}
    </UserContext.Provider>
  );
};

export const useUserContext = () => useContext(UserContext);
