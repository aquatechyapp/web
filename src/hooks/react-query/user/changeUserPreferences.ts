import { useMutation } from '@tanstack/react-query';
import { AxiosError } from 'axios';

import { useToast } from '@/components/ui/use-toast';

import { clientAxios } from '../../../lib/clientAxios';

type UserEmailPreferences = {
  sendEmails: boolean;
  attachChemicalsReadings: boolean;
  attachChecklist: boolean;
  attachServiceNotes: boolean;
  attachServicePhotos: boolean;
  ccEmail: string;
};

export const useChangeUserPreferences = () => {
  const { toast } = useToast();

  const { mutate, isPending, data, isSuccess } = useMutation({
    mutationFn: async (userEmailPreferences: UserEmailPreferences) =>
      await clientAxios.patch('/users/preferences', { serviceEmailPreferences: { ...userEmailPreferences } }),
    onSuccess: () => {
      toast({
        duration: 2000,
        title: 'Email preferences updated successfully',
        variant: 'success'
      });
    },
    onError: (
      error: AxiosError<{
        message: string;
      }>
    ) => {
      toast({
        duration: 2000,
        variant: 'error',
        title: 'Error updating preferences email settings',
        description: error.response?.data?.message ? error.response.data.message : 'Internal server error'
      });
    }
  });
  return { mutate, isPending, data, isSuccess };
};
