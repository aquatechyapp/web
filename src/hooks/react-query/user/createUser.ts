import { useMutation } from '@tanstack/react-query';
import axios, { isAxiosError } from 'axios';
import { z } from 'zod';
import { clientAxios } from '@/lib/clientAxios';
import { useRouter } from 'next/navigation';

const formSchema = z
  .object({
    email: z.string().email(),
    password: z.string().min(8),
    confirmPassword: z.string().min(8)
  })
  .refine(
    (data) => {
      return data.password === data.confirmPassword;
    },
    {
      message: 'Passwords do not match',
      path: ['confirmPassword']
    }
  );

type SignupData = z.infer<typeof formSchema>;

export const useSignup = () => {
  const router = useRouter();

  const { mutate, isPending } = useMutation({
    mutationFn: async (data: SignupData) => {
      try {
        const response = await clientAxios.post('/createuser', data);
        return response.data;
      } catch (error) {
        if (isAxiosError(error) && error.response?.status === 409) {
          throw new Error('User with this e-mail already exists.');
        }
        throw new Error('Error creating user');
      }
    }
  });

  return { signup: mutate, isPending };
};
