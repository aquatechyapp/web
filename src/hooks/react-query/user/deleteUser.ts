import { useMutation } from '@tanstack/react-query';
import { AxiosError } from 'axios';
import { useRouter } from 'next/navigation';

import { toast } from '../../../components/ui/use-toast';
import { clientAxios } from '../../../lib/clientAxios';

export const useDeleteUser = () => {
  const router = useRouter();
  const { mutate, isPending, isError } = useMutation({
    mutationFn: async () => await clientAxios.delete('/users'),
    onSuccess: async () => {
      toast({
        duration: 2000,
        title: 'Your account has been deleted',
        variant: 'success'
      });
      router.push('/login');
    },
    onError: (
      error: AxiosError<{
        message: string;
      }>
    ) => {
      toast({
        duration: 2000,
        variant: 'error',
        title: 'Error deleting your account',
        description: error.response?.data?.message ? error.response.data.message : 'Internal server error'
      });
    }
  });
  return { mutate, isPending, isError };
};
