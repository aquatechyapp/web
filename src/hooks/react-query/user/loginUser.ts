import { useMutation } from '@tanstack/react-query';
import { AxiosError, isAxiosError } from 'axios';
import Cookies from 'js-cookie';
import { useRouter } from 'next/navigation';

import { toast } from '../../../components/ui/use-toast';
import { clientAxios } from '../../../lib/clientAxios';

type LoginData = {
  email: string;
  password: string;
};

export const useLoginUser = () => {
  const { push } = useRouter();

  const { mutate, isPending, error } = useMutation({
    mutationFn: async (data: LoginData) => await clientAxios.post('/sessions', data),
    onSuccess: async ({ data }) => {
      Cookies.set('accessToken', data.accessToken);
      Cookies.set('userId', data.user.id);
      // setUser({
      //   ...data.user,
      //   incomeAsACompany: data.incomeAsACompany,
      //   incomeAsASubcontractor: data.incomeAsASubcontractor
      // });
      if (data.user.firstName === '') {
        push('/account');
        toast({
          duration: 5000,
          title: 'Login successful! Please complete your profile before starting using the platform.',
          variant: 'success'
        });
        return;
      }
      push('/dashboard');
      toast({
        duration: 5000,
        title: 'Login successful!',
        variant: 'success'
      });
    },
    onError: (error): Error | AxiosError => {
      if (isAxiosError(error)) {
        if (error.response?.status === 401 && error.response?.data.message === 'Invalid email or password.') {
          toast({
            variant: 'error',
            duration: 5000,
            title: 'Invalid email or password'
          });
          return error;
        }
      }
      if (error.message === 'Network Error') {
        toast({
          duration: 5000,
          title: 'Internal error',
          variant: 'error'
        });
      }
      return error;
    }
  });
  return { mutate, isPending, error };
};
