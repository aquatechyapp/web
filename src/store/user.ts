import { create } from 'zustand';
import { devtools } from 'zustand/middleware';

import { UserSubscription } from '@/ts/enums/enums';

import { User } from '../ts/interfaces/User';

type Store = {
  user: User;
  isFreePlan: boolean;
  reachedPoolLimit: boolean;
  shouldDisableNewPools: boolean;
};

type Actions = {
  setUser: (user: User) => void;
};

const poolsLimitBySubscription = {
  [UserSubscription.FREE]: 30,
  [UserSubscription.GROW]: 1000
};

export const useUserStore = create<Store & Actions>()(
  devtools((set) => ({
    user: {} as User,
    setUser: (user: User) => {
      set({
        user,
        isFreePlan: user.subscription === UserSubscription.FREE,
        reachedPoolLimit: user.poolsCount >= poolsLimitBySubscription[user.subscription],
        shouldDisableNewPools: user.poolsCount >= poolsLimitBySubscription[user.subscription]
      });
    },
    isFreePlan: false,
    reachedPoolLimit: false,
    shouldDisableNewPools: false
  }))
);
