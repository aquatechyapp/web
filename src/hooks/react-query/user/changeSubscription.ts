import { useMutation } from '@tanstack/react-query';
import { AxiosError } from 'axios';

import { useToast } from '@/components/ui/use-toast';
import { UserSubscription } from '@/constants/enums';
import { useUserStore } from '@/store/user';

import { clientAxios } from '../../../lib/clientAxios';

export const useChangeSubscription = (subscriptionPlan: UserSubscription) => {
  const user = useUserStore((state) => state.user);
  const { toast } = useToast();

  const { mutate, isPending, data, isSuccess } = useMutation({
    mutationFn: async () =>
      await clientAxios.post('/subscriptions', {
        userId: user.id,
        email: user.email,
        firstName: user.firstName,
        lastName: user.lastName,
        company: user.company,
        subscriptionPlan
      }),
    onError: (error): Error | AxiosError => {
      toast({
        title: 'Error on request to change subscription',
        description: 'Please try again later or contact support',
        className: 'bg-red-500 text-white'
      });
      return error;
    }
  });
  return { mutate, isPending, data, isSuccess };
};
