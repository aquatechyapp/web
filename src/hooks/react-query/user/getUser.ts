import { useQuery } from '@tanstack/react-query';
import { useShallow } from 'zustand/react/shallow';

import { clientAxios } from '@/lib/clientAxios';
import { useUserStore } from '@/store/user';
import { SubcontractorStatus } from '@/ts/enums/enums';
import { User } from '@/ts/interfaces/User';
import { useMembersStore } from '@/store/members';

type Props = {
  userId: string;
};

export default function useGetUser({ userId }: Props) {
  const setUser = useUserStore((state) => state.setUser);
  const setDashboard = useUserStore((state) => state.setDashboard);
  const { setAssignmentToId, setAssignedToId } = useMembersStore(
    useShallow((state) => ({
      setMembers: state.setMembers,
      setAssignmentToId: state.setAssignmentToId,
      setAssignedToId: state.setAssignedToid
    }))
  );

  const { data, isLoading, isSuccess } = useQuery({
    queryKey: ['user', userId],
    queryFn: async () => {
      const response = (await clientAxios.get(`/users/${userId}/v2`)).data;
      const user = response.data.user as User;
      delete response.data.user;
      
      const dashboard = response.data.dashboard;

      setUser(user);
      setDashboard(dashboard);
      setAssignmentToId(user.id);
      setAssignedToId(user.id);

      return user;
    }
  });

  return { data, isLoading, isSuccess };
}
