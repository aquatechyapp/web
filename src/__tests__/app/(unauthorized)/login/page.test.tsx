import { render } from '@testing-library/react';

import LoginPage from '@/app/(unauthorized)/login/page';
import { useLoginUser } from '@/hooks/react-query/user/loginUser';

vi.mock('@/hooks/react-query/user/loginUser');

describe('LoginPage', () => {
  vi.mocked(useLoginUser).mockReturnValue({
    mutate: vi.fn(),
    isPending: false,
    error: null
  });
  it('should render LoginPage', () => {
    const wrapper = render(<LoginPage />);
    expect(wrapper).toMatchSnapshot();
  });
});
