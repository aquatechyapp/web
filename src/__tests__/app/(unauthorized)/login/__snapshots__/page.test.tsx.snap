// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > should render LoginPage 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="inline-flex w-[448px] flex-col items-start justify-start gap-[18px] rounded-lg bg-gray-50 px-6 py-8"
      >
        <div
          class="inline-flex h-5 items-center justify-center gap-3 self-stretch"
        >
          <img
            alt="Logo"
            class="h-auto w-80"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="0"
            sizes="100vw"
            src="/_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=3840&q=75"
            srcset="/_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=640&q=75 640w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=750&q=75 750w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=828&q=75 828w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=1080&q=75 1080w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=1200&q=75 1200w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=1920&q=75 1920w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=2048&q=75 2048w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=3840&q=75 3840w"
            style="color: transparent;"
            width="0"
          />
        </div>
        <div
          class="relative mt-4 h-[50px] w-[400px]"
        >
          <div
            class="left-0 top-0 h-[30px] w-[400px] text-xl font-semibold leading-[30px] text-gray-900"
          >
            Login
          </div>
          <div
            class="left-0 top-[30px] h-5 w-[400px]"
          >
            <span
              class="text-sm font-medium text-gray-500"
            >
              Don't you have an account?
            </span>
            <a
              class="text-sm font-semibold text-blue-500"
              href="/signup"
            >
              Sign Up
            </a>
          </div>
        </div>
        <form>
          <div
            class="mb-8 flex w-[400px] flex-col gap-[18px]"
          >
            <div
              class="grid h-full w-full gap-2"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-nowrap"
              >
                E-mail
              </label>
              <input
                aria-describedby=":r0:-form-item-description"
                aria-invalid="false"
                class="flex h-9 w-full rounded-md border border-slate-200 bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-slate-500 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-slate-950 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-800 dark:placeholder:text-slate-400 dark:focus-visible:ring-slate-300"
                id=":r0:-form-item"
                name="email"
                placeholder="E-mail address"
                type="text"
                value=""
              />
            </div>
            <div
              class="grid h-full w-full gap-2"
            >
              <input
                aria-describedby=":r1:-form-item-description"
                aria-invalid="false"
                class="flex h-9 w-full rounded-md border border-slate-200 bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-slate-500 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-slate-950 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-800 dark:placeholder:text-slate-400 dark:focus-visible:ring-slate-300"
                id=":r1:-form-item"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <a
              class="text-sm font-semibold text-blue-500"
              href="/recover"
            >
              Forgot Password?
            </a>
          </div>
          <button
            class="inline-flex items-center justify-center gray-50space-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-slate-950 disabled:pointer-events-none disabled:opacity-50 dark:focus-visible:ring-slate-300 bg-slate-900 text-slate-50 shadow hover:bg-slate-900/90 dark:bg-slate-50 dark:text-slate-900 dark:hover:bg-slate-50/90 h-9 px-4 py-2 w-full"
            type="submit"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="inline-flex w-[448px] flex-col items-start justify-start gap-[18px] rounded-lg bg-gray-50 px-6 py-8"
    >
      <div
        class="inline-flex h-5 items-center justify-center gap-3 self-stretch"
      >
        <img
          alt="Logo"
          class="h-auto w-80"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="0"
          sizes="100vw"
          src="/_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=3840&q=75"
          srcset="/_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=640&q=75 640w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=750&q=75 750w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=828&q=75 828w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=1080&q=75 1080w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=1200&q=75 1200w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=1920&q=75 1920w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=2048&q=75 2048w, /_next/image?url=%2Fpublic%2Fimages%2FlogoHor.png&w=3840&q=75 3840w"
          style="color: transparent;"
          width="0"
        />
      </div>
      <div
        class="relative mt-4 h-[50px] w-[400px]"
      >
        <div
          class="left-0 top-0 h-[30px] w-[400px] text-xl font-semibold leading-[30px] text-gray-900"
        >
          Login
        </div>
        <div
          class="left-0 top-[30px] h-5 w-[400px]"
        >
          <span
            class="text-sm font-medium text-gray-500"
          >
            Don't you have an account?
          </span>
          <a
            class="text-sm font-semibold text-blue-500"
            href="/signup"
          >
            Sign Up
          </a>
        </div>
      </div>
      <form>
        <div
          class="mb-8 flex w-[400px] flex-col gap-[18px]"
        >
          <div
            class="grid h-full w-full gap-2"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-nowrap"
            >
              E-mail
            </label>
            <input
              aria-describedby=":r0:-form-item-description"
              aria-invalid="false"
              class="flex h-9 w-full rounded-md border border-slate-200 bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-slate-500 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-slate-950 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-800 dark:placeholder:text-slate-400 dark:focus-visible:ring-slate-300"
              id=":r0:-form-item"
              name="email"
              placeholder="E-mail address"
              type="text"
              value=""
            />
          </div>
          <div
            class="grid h-full w-full gap-2"
          >
            <input
              aria-describedby=":r1:-form-item-description"
              aria-invalid="false"
              class="flex h-9 w-full rounded-md border border-slate-200 bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-slate-500 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-slate-950 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-800 dark:placeholder:text-slate-400 dark:focus-visible:ring-slate-300"
              id=":r1:-form-item"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <a
            class="text-sm font-semibold text-blue-500"
            href="/recover"
          >
            Forgot Password?
          </a>
        </div>
        <button
          class="inline-flex items-center justify-center gray-50space-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-slate-950 disabled:pointer-events-none disabled:opacity-50 dark:focus-visible:ring-slate-300 bg-slate-900 text-slate-50 shadow hover:bg-slate-900/90 dark:bg-slate-50 dark:text-slate-900 dark:hover:bg-slate-50/90 h-9 px-4 py-2 w-full"
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
